### PR TITLE
Add support for pluggable Docker exec handlers

### PR DESCRIPTION
**Pull request by ncdc**
https://github.com/GoogleCloudPlatform/kubernetes/pull/8864

### CODE REVIEW

**Comment by ncdc** [comment]
[test]

**Comment by ncdc** [comment]
@smarterclayton @danmcp this will switch us from nsenter to Docker's exec API. If you want, I can add a flag to the node startup options to allow the exec handler to be specified (native, nsenter, and later on, @mrunalp's dockerexec), but I'm not sure which node config structs to modify, as I've found 3 or 4.

**Comment by smarterclayton** [comment]
Talk to @deads2k

**Comment by deads2k** [comment]
@ncdc We're trying to keep flags minimal, but have a very expressive config.  This seems like a place where we can choose a reasonable default for the command and the `NodeConfig` (https://github.com/openshift/origin/blob/master/pkg/cmd/server/api/types.go#L10).  That mean you'll need to a field for the option, validation, and defaulting code.

**Comment by liggitt** [comment]
see BuildSerializeableNodeConfig() for how we build the default config from command-line args

to handle setting a default value for configs which don't have one set, see server/api/v1/conversions.go#AddDefaultingFuncs

add validation in server/api/validation/node.go

**Comment by ncdc** [comment]
@deads2k @liggitt thanks guys. It doesn't need to be a flag by any means. I'll look at your pointers and add another commit w/the config setting soon.

**Comment by ncdc** [comment]
Jenkins failure: `fatal: '/data/src/github.com/openshift/origin-bare' does not appear to be a git repository`. So much :heart: :heart: :heart: 

[test]

**Comment by ncdc** [comment]
[test]

**Comment by ncdc** [comment]
[test]

File: pkg/cmd/server/admin/create_nodeconfig.go [hunk 1]
```diff
@@ -395,7 +395,17 @@ func (o CreateNodeConfigOptions) MakeNodeConfig(serverCertFile, serverKeyFile, n
 		return err
 	}
 
-	content, err := latestconfigapi.WriteYAML(config)
+	// Roundtrip the config to v1 and back to ensure proper defaults are set.
```
Comment (ncdc) [inline pkg/cmd/server/admin/create_nodeconfig.go]: @deads2k with @liggitt's help we discovered we need to add this to get defaulting of the NodeConfig. He suggested I let you know as we should probably unify this and BuildSerializeableNodeConfig

**Comment by ncdc** [comment]
Ok, config option is in place. Native (docker exec) is now the default. @smarterclayton PTAL

File: pkg/cmd/server/admin/create_nodeconfig.go (hunk 1 above)
Comment (deads2k) [inline pkg/cmd/server/admin/create_nodeconfig.go]: > He suggested I let you know as we should probably unify this and BuildSerializeableNodeConfig

I agree.  Do we make github issues for tech debt?

Comment (liggitt) [inline pkg/cmd/server/admin/create_nodeconfig.go]: I have been

Comment (deads2k) [inline pkg/cmd/server/admin/create_nodeconfig.go]: https://github.com/openshift/origin/issues/2591  @liggitt please assign to me or to @stevekuznetsov

File: pkg/cmd/server/api/types.go [hunk 2]
```diff
@@ -40,6 +49,10 @@ type NodeConfig struct {
 	// PodManifestConfig holds the configuration for enabling the Kubelet to
 	// create pods based from a manifest file(s) placed locally on the node
 	PodManifestConfig *PodManifestConfig
+
+	// DockerExecHandlerName is the name of the handler to use for executing
+	// commands in Docker containers.
+	DockerExecHandlerName DockerExecHandlerType
```
Comment (deads2k) [inline pkg/cmd/server/api/types.go]: Is there likely to be additional configuration of a handler possible?  If so, a separate stanza is in order.

Comment (liggitt) [inline pkg/cmd/server/api/types.go]: A stanza for docker options would probably be good. If we'd known ahead of time, I'd have put `AllowDisabledDocker` in there as well

Comment (deads2k) [inline pkg/cmd/server/api/types.go]: Conversions!

Comment (ncdc) [inline pkg/cmd/server/api/types.go]: > Conversions!

@deads2k ???

Comment (deads2k) [inline pkg/cmd/server/api/types.go]: > Conversions!
> @deads2k ???

Sorry, I got overly excited at the prospect of carrying conversion functions to move `AllowDisabledDocker` into a `DockerOptions` stanza of our node config.

Let's start `DockerOptions` with the `DockerExecHandlerName` and deal with refactoring other pieces later.

Comment (ncdc) [inline pkg/cmd/server/api/types.go]: Can that wait till later?

Comment (deads2k) [inline pkg/cmd/server/api/types.go]: > Can that wait till later?

Sorry, config lives forever.  We don't have stable and unstable levels like the API.  The longer we wait, the more conversions we have to keep around forever.

I'll take care of moving old values and their associated conversions, so your code can be conversion free: just a new `DockerConfig` struct containing `DockerExecHandlerName`.

**Comment by deads2k** [comment]
config lgtm

**Comment by smarterclayton** [comment]
No issues from me

**Comment by ncdc** [comment]
[merge]

**Pull request merged**